### PR TITLE
A Friendly Beginner's Guide to TypeScript

### LINKED ISSUE #1291: A Friendly Beginner's Guide to TypeScript
### Brief Summary: 
TypeScript is a programming language that extends JavaScript by adding types. It saves you time catching errors and providing fixes before you run code. This tutorial is to provide a beginner guide to a developer trying to upskill or learn TypeScript.

### Key Takeaways: 
- Understand TypeScript fundamentals
- Compiling and Configuration of TypeScript 
- Classes in Typescript
- Interface in TypeScript
- Module and Namespaces

### References: 
https://www.typescriptlang.org/docs/
https://github.com/dzharii/awesome-typescript

### PR DESCRIPTION
**Pull request by Olanetsoft**
#### Preliminary Checks - Formatting and Structure 
- [ ] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [ ] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [ ] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [ ] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [ ] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [ ] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [ ] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [ ] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [ ] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [ ] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [ ] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [ ] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

This Fixes #1291

### CODE REVIEW

File: articles/a-friendly-beginner-guide-to-typescript/index.md [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+### String
+
+TypeScript uses single (') or double(") quote similar to JavaScript.
+
+```
+let colour: string = "yellow";
+color = 'white';
+```
+
+Template Strings is also valid in typeScript and can be used without limitations.
```
Comment (kayere) [inline articles/a-friendly-beginner-guide-to-typescript/index.md:51]: "Template strings are..."

File: articles/a-friendly-beginner-guide-to-typescript/index.md [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+You can manually access set the number of an enum.
+
+```
+enum Fruits {
+  Orange = 1,
+  Apple = 4,
+  Banana= 7
+}
+// can be accessed by
+let c: Fruits = Fruits.Bannana;
```
Comment (kayere) [inline articles/a-friendly-beginner-guide-to-typescript/index.md:82]:
```suggestion
let c: Fruits = Fruits.Banana;
```

File: articles/a-friendly-beginner-guide-to-typescript/index.md [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+```
+
+### Array
+
+TypeScript also supports array similar to javascript and its represented as
```
Comment (kayere) [inline articles/a-friendly-beginner-guide-to-typescript/index.md:87]:
```suggestion
TypeScript also supports arrays. They are represented as;
```

File: articles/a-friendly-beginner-guide-to-typescript/index.md [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+amount = "Hello World";
+amount = [];
+amount = {};
+amount = null;
+amount = undefined;
+```
+
+### Unknown
+
+This works similar to Any but when you try to reassign the value initialized to a new variable then you get an error, Which means anything is assignable to unknown but itself and `any`core types.
```
Comment (kayere) [inline articles/a-friendly-beginner-guide-to-typescript/index.md:121]: This sentence is hard to read. Please rephrase it.

File: articles/a-friendly-beginner-guide-to-typescript/index.md [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+function greetings(): void {
+  console.log("Hi everyone 😊");
+}
+```
+
+You can also check out the [TypeScript Doc](https://www.typescriptlang.org/docs) to learn more about the types and basics in TypeScript.
+
+## Compiler Configurations
+
+tsconfig.json is used to specify the root files and the compiler options required to compile the project whenever `tsc` is run locally. e.g
```
Comment (kayere) [inline articles/a-friendly-beginner-guide-to-typescript/index.md:143]: Add backticks to the file name.

File: articles/a-friendly-beginner-guide-to-typescript/index.md [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+
+### To run your TypeScript File
+
+- Configure typescript using the ` tsconfig.json` file.
+- Run `tsc --watch`, so that every time you change a `.ts` file, tsc will compile it and produce the output in any folder configured e.g a dist folder.
+- Use nodemon to watch if files in ./dist have changed and if needed to relaunch the server.
+
+configure your script in package.json to look like this
+
+```
```
Comment (kayere) [inline articles/a-friendly-beginner-guide-to-typescript/index.md:232]: Add the language name after the opening backticks. Do this for all code snippets.

File: articles/a-friendly-beginner-guide-to-typescript/index.md [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+    "build": "tsc",
+    "watch:build": "tsc --watch",
+    "watch:server": "nodemon './dist/index.js' --watch './dist'",
+    "start": "npm-run-all clean build --parallel watch:build watch:server --print-label"
+  },
+```
+
+## Classes & Interfaces
+
+Interface in TypeScript is a structured group of properties that describe an object.
```
Comment (kayere) [inline articles/a-friendly-beginner-guide-to-typescript/index.md:244]:
```suggestion
An interface is a structured group of properties that describe an object.
```

File: articles/a-friendly-beginner-guide-to-typescript/index.md [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+
+Class is a blueprint from which we can create objects that share the same configuration or properties and methods.
```
Comment (kayere) [inline articles/a-friendly-beginner-guide-to-typescript/index.md:246]:
```suggestion
A class is a blueprint from which we can create objects that share the same configuration or properties and methods.
```

**Review by kayere — COMMENTED** [review]
Hello, I am done with the preliminary review. Please make the suggested changes so that I can proceed to technical review.

File: articles/a-friendly-beginner-guide-to-typescript/index.md (hunk 8 above)
Comment (Olanetsoft) [reply articles/a-friendly-beginner-guide-to-typescript/index.md]: Done

File: articles/a-friendly-beginner-guide-to-typescript/index.md (hunk 7 above)
Comment (Olanetsoft) [reply articles/a-friendly-beginner-guide-to-typescript/index.md]: > 

Done

File: articles/a-friendly-beginner-guide-to-typescript/index.md (hunk 6 above)
Comment (Olanetsoft) [reply articles/a-friendly-beginner-guide-to-typescript/index.md:257]: Done

File: articles/a-friendly-beginner-guide-to-typescript/index.md (hunk 5 above)
Comment (Olanetsoft) [reply articles/a-friendly-beginner-guide-to-typescript/index.md]: Added

File: articles/a-friendly-beginner-guide-to-typescript/index.md (hunk 4 above)
Comment (Olanetsoft) [reply articles/a-friendly-beginner-guide-to-typescript/index.md]: Done

File: articles/a-friendly-beginner-guide-to-typescript/index.md (hunk 3 above)
Comment (Olanetsoft) [reply articles/a-friendly-beginner-guide-to-typescript/index.md]: Fixed

File: articles/a-friendly-beginner-guide-to-typescript/index.md (hunk 2 above)
Comment (Olanetsoft) [reply articles/a-friendly-beginner-guide-to-typescript/index.md]: Done

File: articles/a-friendly-beginner-guide-to-typescript/index.md (hunk 1 above)
Comment (Olanetsoft) [reply articles/a-friendly-beginner-guide-to-typescript/index.md]: Fixed

File: articles/a-friendly-beginner-guide-to-typescript/index.md [hunk 9]
```diff
@@ -0,0 +1,309 @@
+# What is TypeScript?
```
[the remark this answers is not in the document]
Comment (Olanetsoft) [reply articles/a-friendly-beginner-guide-to-typescript/index.md]: Done

**Comment by Olanetsoft** [comment]
updated all changes
@kayere

**Review by kayere — CHANGES_REQUESTED** [review]
Hello @Olanetsoft, I've gone through the tech review of you article and I think that the article needs more information. The article is great but does not give the reader all the relevant details to get started with typescript. Furthermore, the article word count(minus code snippets) is less than 750 words. Please add some more information to make the article more useful to the community.

Lastly, you added a lot of commits when making the changes that I had requested. These commits have corrupted your PR; it now contains too many articles. Please fix that.

**Comment by Olanetsoft** [comment]
> Hello @Olanetsoft, I've gone through the tech review of you article and I think that the article needs more information. The article is great but does not give the reader all the relevant details to get started with typescript. Furthermore, the article word count(minus code snippets) is less than 750 words. Please add some more information to make the article more useful to the community.
> 
> Lastly, you added a lot of commits when making the changes that I had requested. These commits have corrupted your PR; it now contains too many articles. Please fix that. 

@kayere thanks for the feedback, I will update with the requested changes. 

I only made one commit as regard to the changes you requested but I pulled and rebase according to the instruction in the PR guide. 

I will be glad if you can guide me on how to resolve that.

**Comment by kayere** [comment]
> > Hello @Olanetsoft, I've gone through the tech review of you article and I think that the article needs more information. The article is great but does not give the reader all the relevant details to get started with typescript. Furthermore, the article word count(minus code snippets) is less than 750 words. Please add some more information to make the article more useful to the community.
> > Lastly, you added a lot of commits when making the changes that I had requested. These commits have corrupted your PR; it now contains too many articles. Please fix that.
> 
> @kayere thanks for the feedback, I will update with the requested changes.
> 
> I only made one commit as regard to the changes you requested but I pulled and rebase according to the instruction in the PR guide.
> 
> I will be glad if you can guide me on how to resolve that.

Okay. Let me see how that can be fixed. 
Only pull and rebase when creating the pull request.

**Comment by kayere** [comment]
Please don't make any changes until I am done fixing.

**Comment by Olanetsoft** [comment]
@kayere sorry 😞 you mean I should pull and rebase then send a new pull request without the changes you requested ? Afterwards, I should send a commit to implement the changes requested only when you a are through with the fix ?

**Comment by kayere** [comment]
@Olanetsoft - Just do the changes. We will do the fix once the article is ready to be published.

**Comment by Olanetsoft** [comment]
Okay. Thank you very much

**Comment by Olanetsoft** [comment]
@kayere I have updated the changes.

**Comment by kayere** [comment]
@Olanetsoft - did you pull my changes before making yours? The article now appears to have some issues. Some headings are missing and some of your backticks are placed incorrectly. Please revise the article. Make sure to include the changes I made in my `minor` commit.

**Comment by kayere** [comment]
Also, add a hero image

**Comment by Olanetsoft** [comment]
@kayere hero image added and changes updated

**Comment by hectorkambow** [comment]
@Olanetsoft  Hello and welcome again! 👋  

Hector here - I'll be doing your final reviews on the PR - im noticing that there are many commits on this pr (see image)
image
The best way to fix this would be to create a new (clean PR) - the PR should only include the files needed for the articles (min 2)

**Comment by Olanetsoft** [comment]
Okay, thanks for the feedback. I will send a Clean PR shortly

**Comment by Olanetsoft** [comment]
New PR created.

**Pull request closed without merging**